lib/grpc/client/adapters/mint/stream_response_process.ex

```diff
@@ -0,0 +1,178 @@
+defmodule GRPC.Client.Adapters.Mint.StreamResponseProcess do
+  @moduledoc """
+  This module represents the process responsible for consuming the
+    incoming messages from a connection. For each request, there will be
+    a process responsible for consuming its messages. At the end of a stream
+    this process will automatically be killed.
+  """
+
+  @typep accepted_types :: :data | :trailers | :headers | :error
+  @typep data_types :: binary() | Mint.Types.headers() | Mint.Types.error()
+
+  @accepted_types [:data, :trailers, :headers, :error]
+  @header_types [:headers, :trailers]
+
+  use GenServer
+
+  @spec start_link(GRPC.Client.Stream.t(), send_headers_or_trailers? :: boolean()) ::
+          GenServer.on_start()
+  def start_link(stream, send_headers_or_trailers?) do
+    GenServer.start_link(__MODULE__, {stream, send_headers_or_trailers?})
+  end
+
+  @doc """
+  Given a pid from this process, build an Elixir.Stream that will consume the accumulated
+    data inside this process
+  """
+  @spec build_stream(pid()) :: Enumerable.t()
+  def build_stream(pid) do
+    Elixir.Stream.unfold(pid, fn pid ->
+      case GenServer.call(pid, :get_response, :infinity) do
+        nil -> nil
+        response -> {response, pid}
+      end
+    end)
+  end
+
+  @doc """
+  Cast a message to process to inform that the stream has finished
+    once all messages are produced. This process will automatically
+    be killed.
+  """
+  @spec done(pid()) :: :ok
+  def done(pid) do
+    GenServer.cast(pid, {:consume_response, :done})
+  end
+
+  @doc """
+  Consume an incoming data or trailers/headers
+  """
+  @spec consume(pid(), type :: accepted_types, data :: data_types) :: :ok
+  def consume(pid, type, data) when type in @accepted_types do
+    GenServer.cast(pid, {:consume_response, {type, data}})
```

Comment:
Does this _have_ to be a cast? Casts are non-blocking and behave in a fire-and-forget way.

Reply:
Yup! in this case this one makes sense to be a cast. Cause this process will be responsible (and should be the one responsible) for handling errors related to a single request. 

 I pictured this to to keep the connection process responsible for only deal with connection stuff (aka: manage the connection state and properly handle the the low level part of the requests and responses - chunk requests and distribute responses). 

If I use `call` here there's not much I can do to recover from errors from the connection process. Since at the end I designed the adapter to get responses from this module.

Reply:
The problem with cast is that it returns ok even for dead processes. I don't think we should be using it. I'd rather have this be a call even if just for us to send a telemetry event on the failure case

Reply:
ahhh, right. I didn't thought about that. Makes sense. I'll do the change

Reply:
Hey @polvalente just checking this. When you talk about telemetry you're talking about a general telemetry (aka: `Logger.error(...something)` ) or you're thinking about use the telemetry lib to emit such events?

Reply:
It would be ctual telemetry events. But for this PR we can just add a TODO to deal with the failure case there.

The only issue with the current code is that a dead stream process would kill the entire connection